libraries/Networking/src/ESP_Network_Interface.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+
+#include "esp_netif_types.h"
+#include "esp_event.h"
+#include "Arduino.h"
+#include "ESP_Network_Manager.h"
+#include "Printable.h"
+
+typedef enum {
+    ESP_NETIF_ID_STA,
+    ESP_NETIF_ID_AP,
+    ESP_NETIF_ID_PPP,
+    ESP_NETIF_ID_ETH0,
+    ESP_NETIF_ID_ETH1,
+    ESP_NETIF_ID_ETH2,
+    ESP_NETIF_ID_MAX
+} ESP_Network_Interface_ID;
+
+static const int ESP_NETIF_STARTED_BIT          = BIT0;
+static const int ESP_NETIF_CONNECTED_BIT        = BIT1;
+static const int ESP_NETIF_HAS_IP_BIT           = BIT2;
+static const int ESP_NETIF_HAS_LOCAL_IP6_BIT    = BIT3;
+static const int ESP_NETIF_HAS_GLOBAL_IP6_BIT   = BIT4;
+static const int ESP_NETIF_WANT_IP6_BIT         = BIT5;
+
+#define ESP_NETIF_ID_ETH ESP_NETIF_ID_ETH0
+
+class ESP_Network_Interface: public Printable {
+    public:
+        ESP_Network_Interface();
+        virtual ~ESP_Network_Interface();
+
+        // For server interfaces (WiFi AP), dns1 is the DHCP lease range start and dns2 is the DNS. dns3 is not used
+        bool config(IPAddress local_ip = (uint32_t)0x00000000, IPAddress gateway = (uint32_t)0x00000000, IPAddress subnet = (uint32_t)0x00000000, IPAddress dns1 = (uint32_t)0x00000000, IPAddress dns2 = (uint32_t)0x00000000, IPAddress dns3 = (uint32_t)0x00000000);
+        bool dnsIP(uint8_t dns_no, IPAddress ip);
+
+        const char * getHostname() const;
+        bool setHostname(const char * hostname) const;
+
+        bool started() const;
+        bool connected() const;
+        bool hasIP() const;
+        bool hasLocalIPv6() const;
+        bool hasGlobalIPv6() const;
+        bool enableIPv6(bool en=true);
+
+        bool linkUp() const;
+        const char * ifkey() const;
+        const char * desc() const;
+        String impl_name() const;
+
+        uint8_t * macAddress(uint8_t* mac) const;
+        String macAddress() const;
+        IPAddress localIP() const;
+        IPAddress subnetMask() const;
+        IPAddress gatewayIP() const;
+        IPAddress dnsIP(uint8_t dns_no = 0) const;
+        IPAddress broadcastIP() const;
+        IPAddress networkID() const;
+        uint8_t subnetCIDR() const;
+        IPAddress localIPv6() const;
```

Comment:
For link-local scope IPv6 use the name `hasLinkScopeIPv6()` (or `hasLinkLocalScopeIPv6()`). Do not use just the name local, as it is incorrect (it is link-local), and also confusing with term local vs remote, used for connections, and the term localIP() used for IPv4 (which actually usually returns a global scope address).

Local and remote are the different ends of a connection.

Link-local and global are different scopes.

The well established function `localIP()` for IPv4 returns a local (vs remote) address of global scope.

Do not use the same name for IPv6 to mean a different thing.